File: OldBlinks/Blink OverDrive Plus/SDK_new/Blink_SDK_example.m
% Example usage of Blink_SDK_C.dll
% Meadowlark Optics Spatial Light Modulators
% last updated: April 6 2018

% Load the DLL
% Blink_C_wrapper.dll, Blink_SDK.dll, ImageGen.dll, FreeImage.dll and wdapi1021.dll
% should all be located in the same directory as the program referencing the
% library
if ~libisloaded('Blink_C_wrapper')
    loadlibrary('Blink_C_wrapper.dll', 'Blink_C_wrapper.h');
end

% This loads the image generation functions
if ~libisloaded('ImageGen')
    loadlibrary('ImageGen.dll', 'ImageGen.h');
end

% Basic parameters for calling Create_SDK
bit_depth = 12;
num_boards_found = libpointer('uint32Ptr', 0);
constructed_okay = libpointer('int32Ptr', 0);
is_nematic_type = 1;
RAM_write_enable = 1;
use_GPU = 0;
max_transients = 10;
wait_For_Trigger = 0; % This feature is user-settable; use 1 for 'on' or 0 for 'off'
external_Pulse = 0;
timeout_ms = 5000;

% - In your program you should use the path to your custom LUT as opposed to linear LUT
lut_file = 'C:\Program Files\Meadowlark Optics\Blink OverDrive Plus\LUT Files\linear.LUT';
reg_lut = libpointer('string');

% Call the constructor
calllib('Blink_C_wrapper', 'Create_SDK', bit_depth, num_boards_found, constructed_okay, is_nematic_type, RAM_write_enable, use_GPU, max_transients, reg_lut);

% Convention follows that of C function return values: 0 is success, nonzero integer is an error
if constructed_okay.value ~= 0  
    disp('Blink SDK was not successfully constructed');
    disp(calllib('Blink_C_wrapper', 'Get_last_error_message'));
    calllib('Blink_C_wrapper', 'Delete_SDK');
else
    board_number = 1;
    disp('Blink SDK was successfully constructed');
    fprintf('Found %u SLM controller(s)\n', num_boards_found.value);
    
    % load a LUT 
    calllib('Blink_C_wrapper', 'Load_LUT_file',board_number, lut_file);
    
    %allocate arrays for our images
    height = calllib('Blink_C_wrapper', 'Get_image_height', board_number);
    width = calllib('Blink_C_wrapper', 'Get_image_width', board_number);
    ImageOne = libpointer('uint8Ptr', zeros(width*height,1));
    ImageTwo = libpointer('uint8Ptr', zeros(width*height,1));
    WFC = libpointer('uint8Ptr', zeros(width*height,1));
    
    % Generate a blank wavefront correction image, you should load your
    % custom wavefront correction that was shipped with your SLM.
    PixelValue = 0;
    calllib('ImageGen', 'Generate_Solid', WFC, width, height, PixelValue);
    WFC = reshape(WFC.Value, [width,height]);

    % Generate a fresnel lens
    CenterX = width/2;
    CenterY = height/2;
    Radius = height/2;
    Power = 1;
    cylindrical = true;
    horizontal = false;
    calllib('ImageGen', 'Generate_FresnelLens', ImageOne, width, height, CenterX, CenterY, Radius, Power, cylindrical, horizontal);
    ImageOne = reshape(ImageOne.Value, [width,height]);
    ImageOne = rot90(mod(ImageOne + WFC, 256));

    % Generate a blazed grating
    Period = 128;
    Increasing = 1;
    calllib('ImageGen', 'Generate_Grating', ImageTwo, width, height, Period, Increasing, horizontal);
    ImageTwo = reshape(ImageTwo.Value, [width,height]);
    ImageTwo = rot90(mod(ImageTwo + WFC, 256));

      
    % Loop between our two images
    for n = 1:5
	
		%write image returns on DMA complete, ImageWriteComplete returns when the hardware
		%image buffer is ready to receive the next image. Breaking this into two functions is 
		%useful for external triggers. It is safe to apply a trigger when Write_image is complete
		%and it is safe to write a new image when ImageWriteComplete returns
        calllib('Blink_C_wrapper', 'Write_image', board_number, ImageOne, width*height, wait_For_Trigger, external_Pulse, timeout_ms);
		calllib('Blink_C_wrapper', 'ImageWriteComplete', board_number, timeout_ms);
        pause(1.0) % This is in seconds
        calllib('Blink_C_wrapper', 'Write_image', board_number, ImageTwo, width*height, wait_For_Trigger, external_Pulse, timeout_ms);
		calllib('Blink_C_wrapper', 'ImageWriteComplete', board_number, timeout_ms);
        pause(1.0) % This is in seconds
    end
    
    % Always call Delete_SDK before exiting
    calllib('Blink_C_wrapper', 'Delete_SDK');
end

%destruct
if libisloaded('Blink_C_wrapper')
    unloadlibrary('Blink_C_wrapper');
end

if libisloaded('ImageGen')
    unloadlibrary('ImageGen');
end

File: OldBlinks/Blink OverDrive Plus/SDK_new/Blink_SDK_example.cpp
// Blink_SDK_example.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"  // Does nothing but #include targetver.h.

#include "Blink_SDK.h"  // Relative path to SDK header.
#include "ImageGen.h"
#include "math.h"
#include <Windows.h>


// ------------------------- Blink_SDK_example --------------------------------
// Simple example using the Blink_SDK DLL to send a sequence of phase targets
// to a single SLM.
// The code is written with human readability as the main goal.
// The Visual Studio 2013 sample project settings assume that Blink_SDK.lib is
// in relative path ../Blink_SDK/x64/Release.
// To run the example, ensure that Blink_SDK.dll is in the same directory as
// the Blink_SDK_example.exe.
// ----------------------------------------------------------------------------
int main()
{
  int board_number;
  // Construct a Blink_SDK instance with Overdrive capability.
  unsigned int bits_per_pixel = 12U;
  bool         is_nematic_type = true;
  bool         RAM_write_enable = true;
  bool         use_GPU_if_available = true;

  unsigned int n_boards_found = 0U;
  bool         constructed_okay = true;

  Blink_SDK sdk(bits_per_pixel, &n_boards_found,
    &constructed_okay, is_nematic_type, RAM_write_enable,
    use_GPU_if_available, 10U, 0);


  // Check that everything started up successfully.
  bool okay = constructed_okay;

  if (okay)
  {
    board_number = 1;
    //you should replace this with your custom LUT file
	char* lut_file = "C:\\Program Files\\Meadowlark Optics\\Blink OverDrive Plus\\LUT Files\\linear.LUT";
	sdk.Load_LUT_file(board_number, lut_file);

	int height = sdk.Get_image_height(board_number);
	int width = sdk.Get_image_width(board_number);
	// Create two vectors to hold values for two SLM images with opposite ramps.
	unsigned char* ImageOne = new unsigned char[width*height];
	unsigned char* ImageTwo = new unsigned char[width*height];
	// Generate phase gradients
	int VortexCharge = 5;
	Generate_LG(ImageOne, width, height, VortexCharge, width / 2.0, height / 2.0, false);
	VortexCharge = 3;
	Generate_LG(ImageTwo, width, height, VortexCharge, width / 2.0, height / 2.0, false);

	bool ExternalTrigger = false;
	bool OutputPulse = false;
	for (int i = 0; i < 5; i++)
	{
		//write image returns on DMA complete, ImageWriteComplete returns when the hardware
		//image buffer is ready to receive the next image. Breaking this into two functions is 
		//useful for external triggers. It is safe to apply a trigger when Write_image is complete
		//and it is safe to write a new image when ImageWriteComplete returns
		sdk.Write_image(board_number, ImageOne, width*height, ExternalTrigger, OutputPulse, 5000);
		sdk.ImageWriteComplete(board_number, 5000);
		Sleep(500);

		sdk.Write_image(board_number, ImageTwo, width*height, ExternalTrigger, OutputPulse, 5000);
		sdk.ImageWriteComplete(board_number, 5000);
		Sleep(500);
	}

	delete[]ImageOne;
	delete[]ImageTwo;

	sdk.SLM_power(false);
  }

  return (okay) ? EXIT_SUCCESS : EXIT_FAILURE;
}